File: arrays/bribes.py
#!/bin/python3

import math
import os
import random
import re
import sys

from functools import reduce


# Complete the minimumBribes function below.
def minimumBribes(q):
    def f(bribes, enumerable):
        i, starting_place = enumerable
        place_in_line = i + 1

        if starting_place - place_in_line >= 3:
            raise

        best_place_in_line = max(0, starting_place - 2)
        if best_place_in_line < place_in_line:
            bribes_taken = reduce(
                lambda bribes, x: bribes + 1 if x > starting_place else bribes,
                q[best_place_in_line:place_in_line],
                0,
            )
            return bribes + bribes_taken
        else:
            return bribes

    return reduce(f, enumerate(q), 0)


if __name__ == '__main__':
    t = int(input())

    for t_itr in range(t):
        n = int(input())
        q = list(map(int, input().rstrip().split()))
        try:
            print(minimumBribes(q))
        except:
            print("Too chaotic")

File: arrays/bribes.ml
(* Enter your code here. Read input from STDIN. Print output to STDOUT *)
open Core

module Io_util = struct
  let read_n_lines ~n =
    List.init n ~f:(fun _ ->
        let line = In_channel.(input_line stdin) in
        Option.value_exn line)
    |> List.rev
  ;;
end

let handle_case one_case =
  let one_case = List.to_array one_case in
  Array.foldi one_case ~init:0 ~f:(fun i bribes starting_place ->
      let place_in_line = i + 2 in
      if starting_place - place_in_line >= 3 then failwith "Too chaotic";
      match max 1 (starting_place - 2) with
      | best_place_in_line when best_place_in_line < place_in_line ->
        let bribes_taken =
          let best_place_in_line_index = best_place_in_line - 1 in
          Array.slice one_case best_place_in_line_index place_in_line
          |> Array.fold ~init:0 ~f:(fun bribes x ->
                 printf "moo";
                 if x > starting_place then bribes + 1 else bribes)
        in
        bribes_taken + bribes
      | _ -> bribes)
;;

let calculate_bribes () =
  let input_cases = Io_util.read_n_lines ~n:1 |> List.hd_exn |> Int.of_string in
  let test_cases =
    Array.init input_cases ~f:(fun _ ->
        Io_util.read_n_lines ~n:2
        |> List.last_exn
        |> String.split ~on:' '
        |> List.map ~f:Int.of_string)
  in
  Array.iter test_cases ~f:(fun one_case ->
      Option.try_with (fun () -> handle_case one_case)
      |> Option.value_map ~f:Int.to_string ~default:"Too chaotic"
      |> print_endline)
;;

let command = Common.Command_line.create () ~f:calculate_bribes
